### DISCUSSION TITLE
maximum value for "keep_alive_count_max"

### DISCUSSION THREAD

**Discussion by redpimpernel** [gravitational/teleport#31016] [Q&A] [answered]
Hi, does anyone know what the maximum value is for the "keep_alive_count_max" entry in the config?
I am using teleport to access a couple of devices that suffer regular power outages, sometimes for a day.
I know it's not best practice for security to have a large value, but it is something I think I will need.

I have kept the interval at "5m" but upped the count_max vallue to 7200 (5 days) just to see.

I then broke the teleport config on a device I would like to access to see if it would stay listed in teleport. Not sure when, but within a couple of hours it had dropped off and I will need to readd it to the cluster.

Any ideas or suggestions would be great.
Cheers

**Comment by webvictim** [comment, marked as answer]
Extending `keep_alive_count_max` won't actually prevent a node that doesn't send heartbeats from expiring in the backend. It's a setting that controls when incoming connections are considered "lost" and torn down by the backend, but it doesn't override the backend's logic which will prune any host past its expiry time.

If you want to see nodes live longer in the backend, you'd be better off modifying Teleport's heartbeat logic to set an `expires` timestamp way further in the future so that nodes expire less often. This isn't exposed via a config parameter unfortunately, so would require you to fork the repo, change the code and recompile Teleport.

Also, another point of clarification - if a node joins the cluster successfully but subsequently disappears due to flaky connectivity, it will heartbeat again and reappear without any action when the connection is restored. There's no need to rejoin a node that's gone offline unless you perform a CA rotation or similar and the node is offline for the full grace period.

If you're seeing different behaviour, please share logs from the node and Teleport cluster - we'll get to the bottom of it.

**Reply by redpimpernel** [reply]
Hi, Thanks for that detailed explanation.
If I could get you to clear a few things up, that would be great.
You say *It's a setting that controls when incoming connections are considered "lost"*. I am having trouble understanding what that means. If the connections are incoming, how can they be lost?

I feel like it hasn't come back online, could be that I have been impatient and assumed that I needed to generate another join token. What sort of time-frame should I be expecting for a node to appear back in the cluster when connectivity is regained.

Cheers

**Reply by webvictim** [reply]
The behaviour of `keep_alive_interval` and `keep_alive_count_max` actually mimics that of `ServerAliveInterval` and `ServerAliveCountMax` in SSH, so it's easier to read about how it works in this extract from the SSH man page with an explanation: https://serverfault.com/a/538919

Assuming you're running a v13 version of Teleport on the node itself and that remote Teleport process is still running, I'd say you should expect a node to reconnect to the cluster and heartbeat again within a few minutes at the most. There is an exponential backoff inside Teleport on connection attempts, but AFAIK the highest it goes is around 60 seconds.

If you can somehow get and share logs from the node after connectivity is restored, we might be able to offer an explanation as to what's happening.
